pages/blog/graphql-error-handling-with-fp.mdx

```diff
@@ -0,0 +1,631 @@
+import withArticle from '../../ui/blog/article'
+
+export const meta = {
+  title: 'GraphQL error handling to the max with Typescript, codegen and fp-ts',
+  author: 'Ghislain Thau',
+  tags: ['graphql', 'typescript', 'codegen', 'fp-ts'],
+  date: '2022-02-17',
+  description:
+    'Handle unsafe APIs in a typesafe way with Typescript and Functional Programming',
+}
+
+export default withArticle({ ...meta })
+
+Contrary to REST, GraphQL does not use Status Code to differentiate a successful result from an exception. In REST, one can send a response with a appropriate status code to inform the caller about the kind of exception encountered, e.g.:
+- 400: bad request (e.g. invalid inputs)
+- 401 or 403: unauthorized (need authentication) or forbidden (insufficient permission)
+- 404: resource not found
+- 500: internal server error
+- and more.
+
+
+# GraphQL default error handling
+
+
+In GraphQL, you'll get a 5xx error in case the server is not available at all, but for other exceptions you'll get the standard error handling mechanism from GraphQL: in the JSON you receive there is:
+- a `data` key which contains the data corresponding to the GraphQL operation (query, mutation, subscription) invoked and
+- an `errors` key which contains an array of errors returned by the server, with a message and location.
+
+This solution is however not satisfactory for the following reasons:
+
+1. it is not easily parseable, it is more meant for a human user to read
+2. it is not typed: you can add more information, e.g. an error code and other fields, but those are not part of the schema
+3. it is not self-documented: by looking at the operation signature, you don't know what might fail
+4. it treats some domain results as errors / exceptions whereas they are just alternate results of the operation. E.g. an entity not found for a given id and a user not having permission to access an entity belong to the domain model, whereas runtime exceptions such as the GraphQL server failing to communicate with the database / external microservice, or an operation timing out because of slowness are real exceptions: we don't get any result that relates to domain model, just an unexpected runtime error.
+
+
+# Better modelisation of errors using Union and Interfaces
+
+
+GraphQL has support for Union types (only for Types, not yet for Inputs). We can take advantage of it and design our schema to specify the happy and unhappy paths.
```

Comment:
What do "happy and unhappy paths" mean?
Is there a clearer expression we could use? (ex: "expected and erroneous results")

Reply:
Happy path = the expected result. Designing for the happy path is living in a world where nothing unexpected will happen, and obviously will lead to runtime errors.
The "unhappy path" are the alternative paths that exist in addition to the expected result (happy path).
It's quite common to come across this term, e.g. https://en.wikipedia.org/wiki/Happy_path#Unhappy_path, https://cucumber.io/blog/test-automation/happy-unhappy-paths-why-you-need-to-test-both/

I can reword it. We could say "the expected and alternative results"

Reply:
Okay, I see!

I feel like it's a vocabulary mainly from the TDD/BDD world that most people won't get properly (a bit too theoretical)

Rewording it would make it easier to follow for a broader audience 🤓